Update clipboard-win to v5 and replace winapi with windows-sys

rhysd opened the pull request:
This PR does two things for Windows support:

1. Update clipboard-win to v5.0.0
   - I checked [changes in v5.0.0](https://github.com/DoumanAsh/clipboard-win/compare/44487b19cd12a9242f5b1991c088bf81c780f172...973219a363729dcaeb98ba74280f3543322f37d7) and the changes didn't affect any implementation in this crate
3. Replace `winapi` crate dependency with `windows-rs`
   - winapi is unmaintained. windows-rs is well-maintained by Microsoft officially
   - By this change, winapi can be removed completely from this crate
   - The replacement is somewhat mechanical translation from winapi to windows-rs. I read MSDN documents but my understanding of the Win32 APIs is not perfect. I'd like to have a detailed code review
   - Some constants used by this crate are not defined in windows-rs. I defined them directly in `platform/windows.rs`

I confirmed writing/reading clipboard with both text and image contents using `hello_world` and `set_image`/`get_image` examples on my Windows 10 machine.

With this PR, cargo dependencies will change to:

```console
> cargo tree --edges no-dev
arboard v3.3.0
├── clipboard-win v5.0.0
│   └── error-code v3.0.0
├── log v0.4.20
├── thiserror v1.0.35
│   └── thiserror-impl v1.0.35 (proc-macro)
│       ├── proc-macro2 v1.0.43
│       │   └── unicode-ident v1.0.4
│       ├── quote v1.0.21
│       │   └── proc-macro2 v1.0.43 (*)
│       └── syn v1.0.100
│           ├── proc-macro2 v1.0.43 (*)
│           ├── quote v1.0.21 (*)
│           └── unicode-ident v1.0.4
└── windows v0.52.0
    ├── windows-core v0.52.0
    │   └── windows-targets v0.52.0
    │       └── windows_x86_64_msvc v0.52.0
    └── windows-targets v0.52.0 (*)
```

complexspaces commented:
Thanks for the PR, I appreciate the effort. Would you consider modifying this to use `windows-sys` instead of the `windows` crate? The former has much less version churn and I'm not sure how much value `arboard` gets out of using `windows` because we don't utilize any COM or WinRT logic.

I would definitely be in favor of implementing a helper function to wrap some of the `Global*` calls to handle checking for errors internally though.

rhysd commented:
It makes sense. I'll try windows-sys.

rhysd commented:
> I would definitely be in favor of implementing a helper function to wrap some of the Global* calls to handle checking for errors internally though.

I focused on replacing winapi with windows-rs with as small diff as possible. But I'll refactor the code more idiomtic.

complexspaces commented:
No worries either way, I could refactor it later if you are busy. I generally don't mind a little bit of scope creep in PRs that touches the same area as long as its independently reviewable (its own commit, etc).

rhysd commented:
No, problem. I like refactorying so I'll try it by myself :)

rhysd commented on src/platform/windows.rs:
```diff
@@ -8,345 +8,396 @@ the Apache 2.0 or the MIT license at the licensee's choice. The terms
-			CreateDIBitmap, DeleteObject, GetDIBits, LCS_sRGB, BITMAPINFO, BITMAPINFOHEADER,
-			BITMAPV5HEADER, BI_RGB, CBM_INIT, DIB_RGB_COLORS, LCS_GM_IMAGES, PROFILE_EMBEDDED,
-			PROFILE_LINKED, RGBQUAD,
+mod image_data {
```
Separating this module as a new file may be a good idea.

rhysd commented:
I added some refactorings:

- 6cfde9c39e8c0071d52ec0862614af7b14433b16 : Add `Error::unknown` helper function to avoid `Error::Unknown { ... }` boilerplate
- ec5e1d86eaaeec8e6894305949cacb3353f335bb : Separate the implementation details of image data handling into `image_data` module to avoid many repeats of `#[cfg(feature = "image-data")]` gates. Moving the module to a new file may be a good idea (I don't have a strong opinion here).
- ba5575bf843f57463874c0289fd82a6db909229c : Wrap Win32 functions which can return errors to leverage error handling with `?` operator. This made `add_cf_dibv5` and `read_cf_dibv5` simpler
- ba9e7e84c89f880ac930049155b325267aa3156a : Move tests for image data handling to `image_data` module and added more test cases

rhysd commented:
The dependency tree is now:

```console
> cargo tree --edges no-dev
arboard v3.3.0
├── clipboard-win v5.0.0
│   └── error-code v3.0.0
├── log v0.4.20
├── thiserror v1.0.35
│   └── thiserror-impl v1.0.35 (proc-macro)
│       ├── proc-macro2 v1.0.43
│       │   └── unicode-ident v1.0.4
│       ├── quote v1.0.21
│       │   └── proc-macro2 v1.0.43 (*)
│       └── syn v1.0.100
│           ├── proc-macro2 v1.0.43 (*)
│           ├── quote v1.0.21 (*)
│           └── unicode-ident v1.0.4
└── windows-sys v0.52.0
    └── windows-targets v0.52.0
        └── windows_x86_64_msvc v0.52.0
```

complexspaces commented on src/platform/windows.rs:
```diff
@@ -8,345 +8,396 @@ the Apache 2.0 or the MIT license at the licensee's choice. The terms
+		}
+	}
+
+	unsafe fn get_screen_device_context() -> Result<HDC, Error> {
```
Suggestion: This doesn't need to be `unsafe` because `GetDc` doesn't have any safety conditions to uphold beyond the parameter?

rhysd replied:
Yeah, it is actually safe. I made this function safe with a `SAFETY:` comment.

complexspaces commented on src/platform/windows.rs:
```diff
@@ -8,345 +8,396 @@ the Apache 2.0 or the MIT license at the licensee's choice. The terms
-		let result =
-			ImageData { bytes: Cow::Owned(result_bytes), width: w as usize, height: h as usize };
-		Ok(result)
+	/// Copies the bitmap image into given buffer with DIB RGB format and
+	/// returns value is the number of scan lines copied from the bitmap.
```
```suggestion
	/// Copies the bitmap image into given buffer with DIB RGB format and
	/// returns the number of scan lines copied from the bitmap.
```

rhysd commented:
Thank you for your reviews. I'll have a look at them tonight.

rhysd commented on src/platform/windows.rs:
```diff
@@ -8,345 +8,396 @@ the Apache 2.0 or the MIT license at the licensee's choice. The terms
-					line!()
-				),
-			});
+	unsafe fn set_clipboard_data(hmem: HANDLE) -> Result<(), Error> {
```
I created a separate function here because it returns an error and we can levarate `?` error handling to reduce size of big function `add_cf_dibv5`. I agree with the name is not good.

rhysd commented on src/platform/windows.rs:
```diff
@@ -8,345 +8,396 @@ the Apache 2.0 or the MIT license at the licensee's choice. The terms
-use crate::common::{private, Error};
+	fn last_error(message: &str) -> Error {
+		let code = unsafe { GetLastError() };
+		Error::unknown(format!("{} with error code {:x}", message, code))
```
I didn't know the function. It's definitely better.

rhysd commented on src/platform/windows.rs:
```diff
@@ -8,345 +8,396 @@ the Apache 2.0 or the MIT license at the licensee's choice. The terms
+		unsafe {
+			let data_ptr = global_lock(hdata)?;
+			let _unlock = ScopeGuard::new(|| {
+				if GlobalUnlock(hdata) == 0 {
```
Ahh, I misread the MSDN document. I think I overlooked the error log on checking the examples.

complexspaces replied:
I'm still not sure here yet, sorry.

This is the MSDN snippet I've been reading:
> If the memory object is still locked after decrementing the lock count, the return value is a nonzero value. If the memory object is unlocked after decrementing the lock count, the function returns zero and [GetLastError](https://learn.microsoft.com/en-us/windows/desktop/api/errhandlingapi/nf-errhandlingapi-getlasterror) returns NO_ERROR.
>
> If the function fails, the return value is zero and [GetLastError](https://learn.microsoft.com/en-us/windows/desktop/api/errhandlingapi/nf-errhandlingapi-getlasterror) returns a value other than NO_ERROR.

AFAICT, an real error that we would need to log comes from both:
- `GlobalUnlock == 0`
- `LastError != 0`

The `-1` value that is being checked for now doesn't seem to be an error, and is instead just a valid result of calling the function.

rhysd replied:
Ohh, thank you for the point. Since I don't have time to look into this for now, I'll take a look on next Monday.

rhysd commented:
@complexspaces I replied to your review comments and fixed all points. Would you take a look again?

complexspaces reviewed (approved):
Thanks a lot for your work on this, everything looks good now!
